File: app/src/main/java/com/example/rednotedemo/data/paging/PostPagingSource.java
package com.example.rednotedemo.data.paging;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.paging.PagingState;
import androidx.paging.rxjava2.RxPagingSource;

import com.example.rednotedemo.data.repository.PostPageRepository;
import com.example.rednotedemo.entity.vo.PostListItemVO;
import com.example.rednotedemo.common.enums.FilterType;

import org.jetbrains.annotations.NotNull;

import java.util.List;

import io.reactivex.Single;
import io.reactivex.schedulers.Schedulers;

public class PostPagingSource extends RxPagingSource<Integer, PostListItemVO> {
    private PostPageRepository repository;
    private FilterType filterType;

    public PostPagingSource(FilterType filterType, PostPageRepository repository) {
        this.filterType = filterType;
        this.repository = repository;
    }

    @Override
    public @NotNull Single<LoadResult<Integer, PostListItemVO>> loadSingle(@NotNull LoadParams<Integer> loadParams) {
        Integer nextPageNumber = loadParams.getKey();
        if (nextPageNumber == null) nextPageNumber = 1;

        int loadSize = loadParams.getLoadSize(); // 每页条数

        Integer finalNextPageNumber = nextPageNumber;
        return Single.fromCallable(() ->
                        repository.getNeedPagingList(filterType, loadSize, loadSize * (finalNextPageNumber - 1))
                )
                .subscribeOn(Schedulers.io())
                .map(mBeans -> toLoadResult(mBeans, finalNextPageNumber))
                .onErrorReturn(throwable -> new LoadResult.Error<>(throwable));
    }

    private LoadResult<Integer, PostListItemVO> toLoadResult(@NonNull List<PostListItemVO> mBeans, Integer page) {
        Integer prevKey = (page == 1) ? null : page - 1;
        Integer nextKey = mBeans.isEmpty() ? null : page + 1;
        return new LoadResult.Page<>(
                mBeans,
                prevKey,
                nextKey,
                LoadResult.Page.COUNT_UNDEFINED,
                LoadResult.Page.COUNT_UNDEFINED
        );
    }

    @Nullable
    @Override
    public Integer getRefreshKey(@NonNull PagingState<Integer, PostListItemVO> pagingState) {
        return 1;
    }
}

File: app/src/main/java/com/example/rednotedemo/data/paging/PostPagingSource.kt
package com.example.rednotedemo.data.paging

import android.widget.Toast
import androidx.paging.PagingSource
import androidx.paging.PagingState
import com.example.rednotedemo.entity.vo.PostListItemVO
import com.example.rednotedemo.enums.FilterType
import kotlin.random.Random

class PostPagingSource(private val filterType: FilterType) : PagingSource<Int, PostListItemVO>() {

    override suspend fun load(params: LoadParams<Int>): LoadResult<Int, PostListItemVO> {
        try {
            val page = params.key ?: 0
            val pageSize = params.loadSize

            // 模拟数据
            val items = ArrayList<PostListItemVO>()
            for (i in 1..pageSize) {
                val id = Random.nextInt(100000) + page * pageSize + i
                items.add(
                    PostListItemVO()
                        .setPostId(id)
                        .setCoverUrl("https://picsum.photos/300/400?random=$id")
                        .setTitle("动态 $id")
                        .setAuthorName("用户$id")
                        .setLikesCount(10 + id % 100)
                        .setType(filterType)
                )
            }
            return LoadResult.Page(
                data = items,
                prevKey = if (page == 0) null else page - 1,
                nextKey = page + 1
            )

        } catch (e: Exception) {
            return LoadResult.Error(e)
        }
    }

    override fun getRefreshKey(state: PagingState<Int, PostListItemVO>): Int? {
        return state.anchorPosition?.let { anchorPosition ->
            state.closestPageToPosition(anchorPosition)?.prevKey?.plus(1)
                ?: state.closestPageToPosition(anchorPosition)?.nextKey?.minus(1)
        }
    }
}